Fix BPF CT_REPORT_FLAGS variable size

Cilium master (v1.7 cycle)

I noticed these warnings while compiling BPF programs:

```
level=warning msg="+ clang -O2 -g -target bpf -emit-llvm -Wno-address-of-packed-member -Wno-unknown-warning-option -I. -I/run/cilium/state/globals -I/var/lib/cilium/bpf -I/var/lib/cilium/bpf/include -D__NR_CPUS__=4 -DENABLE_ARP_RESP
ONDER -DHANDLE_NS -DSECLABEL=2 -DPOLICY_MAP=cilium_policy_reserved_2 -DBPF_PKT_DIR=1 '-DNODE_MAC={.addr={0x2a,0xcb,0x3f,0xba,0x49,0x0c}}' -DCALLS_MAP=cilium_calls_overlay_2 -c /var/lib/cilium/bpf/bpf_overlay.c -o -" subsys=daemon
level=warning msg="In file included from /var/lib/cilium/bpf/bpf_overlay.c:39:" subsys=daemon                      
level=warning msg="In file included from /var/lib/cilium/bpf/lib/nodeport.h:24:" subsys=daemon                                         
level=warning msg="In file included from /var/lib/cilium/bpf/lib/nat.h:31:" subsys=daemon                                                                                                                                               level=warning msg="/var/lib/cilium/bpf/lib/conntrack.h:178:8: warning: implicit conversion from 'int' to 'uint8_t' (aka 'unsigned char') changes value from 512 to 0 [-Wconstant-conversion]" subsys=daemon                            
level=warning msg="                                   CT_REPORT_FLAGS);" subsys=daemon                         
level=warning msg="                                   ^~~~~~~~~~~~~~~" subsys=daemon                                                                                                                               
level=warning msg="/run/cilium/state/globals/node_config.h:23:25: note: expanded from macro 'CT_REPORT_FLAGS'" subsys=daemon                                                                                                           
level=warning msg="#define CT_REPORT_FLAGS 0x0200" subsys=daemon                      
level=warning msg="                        ^~~~~~" subsys=daemon                                                   
level=warning msg="In file included from /var/lib/cilium/bpf/bpf_overlay.c:39:" subsys=daemon                                                                                                                                          
level=warning msg="In file included from /var/lib/cilium/bpf/lib/nodeport.h:24:" subsys=daemon
level=warning msg="In file included from /var/lib/cilium/bpf/lib/nat.h:31:" subsys=daemon                                              
level=warning msg="/var/lib/cilium/bpf/lib/conntrack.h:272:22: warning: implicit conversion from 'int' to 'uint8_t' (aka 'unsigned char') changes value from 512 to 0 [-Wconstant-conversion]" subsys=daemon
level=warning msg="                                            seen_flags, CT_REPORT_FLAGS);" subsys=daemon                                                                                                    
level=warning msg="                                                        ^~~~~~~~~~~~~~~" subsys=daemon                                                                                                                              
level=warning msg="/run/cilium/state/globals/node_config.h:23:25: note: expanded from macro 'CT_REPORT_FLAGS'" subsys=daemon                                                                                                     
level=warning msg="#define CT_REPORT_FLAGS 0x0200" subsys=daemon                                              
level=warning msg="                        ^~~~~~" subsys=daemon                                                                                                                                                                       
level=warning msg="2 warnings generated." subsys=daemon
```

I have the following option in my configmap:
```
monitor-aggregation-flags: syn
```

--- a/pkg/option/config.go
+++ b/pkg/option/config.go
@@ -29,7 +29,6 @@ import (
 
 	"github.com/cilium/cilium/api/v1/models"
 	"github.com/cilium/cilium/common"
-	"github.com/cilium/cilium/pkg/byteorder"
 	"github.com/cilium/cilium/pkg/cidr"
 	"github.com/cilium/cilium/pkg/defaults"
 	"github.com/cilium/cilium/pkg/ip"
@@ -1787,7 +1786,7 @@ func (c *DaemonConfig) Populate() {
 		}
 		ctMonitorReportFlags |= flag
 	}
-	c.MonitorAggregationFlags = byteorder.HostToNetwork(ctMonitorReportFlags).(uint16)
+	c.MonitorAggregationFlags = ctMonitorReportFlags
 
 	// Map options
 	if m := viper.GetStringMapString(AwsInstanceLimitMapping); len(m) != 0 {

--- a/test/ginkgo-ext/scopes.go
+++ b/test/ginkgo-ext/scopes.go
@@ -250,6 +250,17 @@ func runAllAfterEach(cs *scope, testName string) {
 // already executed for the given test.
 var afterFailedStatus map[string]bool = map[string]bool{}
 
+func testFailed(testName string) bool {
+	hasFailed, _ := afterEachFailed[testName]
+	return ginkgo.CurrentGinkgoTestDescription().Failed || hasFailed
+}
+
+// TestFailed returns true if the current test has failed.
+func TestFailed() bool {
+	testName := ginkgo.CurrentGinkgoTestDescription().FullTestText
+	return testFailed(testName)
+}
+
 // runAllAfterFail runs all the afterFail functions for the given
 // scope and parent scopes. This function make sure that all the `AfterFail`
 // functions are called before AfterEach.
@@ -259,8 +270,7 @@ func runAllAfterFail(cs *scope, testName string) {
 		return
 	}
 
-	hasFailed, _ := afterEachFailed[testName]
-	if (ginkgo.CurrentGinkgoTestDescription().Failed || hasFailed) && len(cs.afterFail) > 0 {
+	if testFailed(testName) && len(cs.afterFail) > 0 {
 		GinkgoPrint("===================== TEST FAILED =====================")
 		for _, body := range cs.afterFail {
 			body()

--- a/test/k8sT/DatapathConfiguration.go
+++ b/test/k8sT/DatapathConfiguration.go
@@ -126,6 +126,32 @@ var _ = Describe("K8sDatapathConfig", func() {
 			ingressPktCount := 2
 			checkMonitorOutput(monitorOutput, egressPktCount, ingressPktCount)
 		})
+
+		It("Checks that monitor aggregation flags send notifications", func() {
+			deployCilium([]string{
+				"--set global.bpf.monitorAggregation=medium",
+				"--set global.bpf.monitorInterval=60s",
+				"--set global.bpf.monitorFlags=psh",
+				"--set global.debug.enabled=false",
+			})
+			monitorOutput, _ := monitorConnectivityAcrossNodes(kubectl, monitorLog)
+
+			By("Checking the set of TCP notifications received matches expectations")
+			// | TCP Flags | Direction | Report? | Why?
+			// +===========+===========+=========+=====
+			// | SYN       |    ->     |    Y    | monitorAggregation=medium
+			// | SYN / ACK |    <-     |    Y    | monitorAggregation=medium
+			// | ACK       |    ->     |    N    | monitorFlags=(!ACK)
+			// | ACK       |    ...    |    N    | monitorFlags=(!ACK)
+			// | PSH       |    ->     |    Y    | monitorFlags=(PSH)
+			// | PSH       |    <-     |    Y    | monitorFlags=(PSH)
+			// | FIN       |    ->     |    Y    | monitorAggregation=medium
+			// | FIN / ACK |    <-     |    Y    | monitorAggregation=medium
+			// | ACK       |    ->     |    Y    | monitorAggregation=medium
+			egressPktCount := 4
+			ingressPktCount := 3
+			checkMonitorOutput(monitorOutput, egressPktCount, ingressPktCount)
+		})
 	})
 
 	Context("Encapsulation", func() {

--- a/test/k8sT/DatapathConfiguration.go
+++ b/test/k8sT/DatapathConfiguration.go
@@ -98,26 +98,7 @@ var _ = Describe("K8sDatapathConfig", func() {
 				"--set global.bpf.monitorFlags=syn",
 				"--set global.debug.enabled=false",
 			})
-			// For local single-node testing, configure
-			// requireMultiNode to false and specify the node name.
-			requireMultiNode := true
-			nodeName := helpers.K8s1
-
-			var err error
-			ciliumPodK8s1, err := kubectl.GetCiliumPodOnNodeWithLabel(helpers.KubeSystemNamespace, nodeName)
-			Expect(err).Should(BeNil(), "Cannot get cilium pod on k8s1")
-
-			By(fmt.Sprintf("Launching cilium monitor on %q", ciliumPodK8s1))
-			monitorStop := kubectl.MonitorStart(helpers.KubeSystemNamespace, ciliumPodK8s1, monitorLog)
-			result, targetIP := testPodConnectivityAcrossNodesAndReturnIP(kubectl, requireMultiNode, 1)
-			monitorStop()
-			cleanService()
-			Expect(result).Should(BeTrue(), "Connectivity test between nodes failed")
-
-			monitorPath := fmt.Sprintf("%s/%s", helpers.ReportDirectoryPath(), monitorLog)
-			By("Reading the monitor log at %s", monitorPath)
-			monitorOutput, err := ioutil.ReadFile(monitorPath)
-			Expect(err).To(BeNil(), "Could not read monitor log")
+			monitorOutput, targetIP := monitorConnectivityAcrossNodes(kubectl, monitorLog)
 
 			By("Checking that exactly one ICMP notification in each direction was observed")
 			expEgress := fmt.Sprintf("ICMPv4.*DstIP=%s", targetIP)
@@ -141,31 +122,9 @@ var _ = Describe("K8sDatapathConfig", func() {
 			// | FIN       |    ->     |    Y    | monitorAggregation=medium
 			// | FIN / ACK |    <-     |    Y    | monitorAggregation=medium
 			// | ACK       |    ->     |    Y    | monitorAggregation=medium
-
-			// Multiple connection attempts may be made, we need to
-			// narrow down to the last connection close, then match
-			// the ephemeral port + flags to ensure that the
-			// notifications match the table above.
-			egressTCPExpr := `TCP.*DstPort=80.*FIN=true`
-			egressTCPRegex := regexp.MustCompile(egressTCPExpr)
-			egressTCPMatches := egressTCPRegex.FindAll(monitorOutput, -1)
-			Expect(len(egressTCPMatches)).To(BeNumerically(">", 0), "Could not locate final FIN notification in monitor log")
-			finalMatch := egressTCPMatches[len(egressTCPMatches)-1]
-			portRegex := regexp.MustCompile(`SrcPort=([0-9]*)`)
-			// FindSubmatch should return ["SrcPort=12345" "12345"]
-			portBytes := portRegex.FindSubmatch(finalMatch)[1]
-
-			By("Looking for TCP notifications using the ephemeral port %q", portBytes)
-			port, err := strconv.Atoi(string(portBytes))
-			Expect(err).To(BeNil(), fmt.Sprintf("ephemeral port %q could not be converted to integer", string(portBytes)))
-			expEgress = fmt.Sprintf("SrcPort=%d", port)
-			expEgressRegex = regexp.MustCompile(expEgress)
-			egressMatches = expEgressRegex.FindAllIndex(monitorOutput, -1)
-			Expect(len(egressMatches)).To(Equal(3), "Monitor log contained unexpected number of egress notifications matching %q", expEgress)
-			expIngress = fmt.Sprintf("DstPort=%d", port)
-			expIngressRegex = regexp.MustCompile(expIngress)
-			ingressMatches = expIngressRegex.FindAllIndex(monitorOutput, -1)
-			Expect(len(ingressMatches)).To(Equal(2), "Monitor log contained unexpected number of ingress notifications matching %q", expIngress)
+			egressPktCount := 3
+			ingressPktCount := 2
+			checkMonitorOutput(monitorOutput, egressPktCount, ingressPktCount)
 		})
 	})
 
@@ -360,4 +319,52 @@ func testPodConnectivityAcrossNodesAndReturnIP(kubectl *helpers.Kubectl, require
 	res = kubectl.ExecPodCmd(helpers.DefaultNamespace, srcPod,
 		helpers.CurlFail("http://%s:80/", targetIP))
 	return res.WasSuccessful(), targetIP
+}
+
+func monitorConnectivityAcrossNodes(kubectl *helpers.Kubectl, monitorLog string) (monitorOutput []byte, targetIP string) {
+	// For local single-node testing, configure requireMultiNode to "false"
+	// and add the labels "cilium.io/ci-node: k8s1" to the node.
+	requireMultiNode := true
+
+	ciliumPodK8s1, err := kubectl.GetCiliumPodOnNodeWithLabel(helpers.KubeSystemNamespace, helpers.K8s1)
+	ExpectWithOffset(1, err).Should(BeNil(), "Cannot get cilium pod on k8s1")
+
+	By(fmt.Sprintf("Launching cilium monitor on %q", ciliumPodK8s1))
+	monitorStop := kubectl.MonitorStart(helpers.KubeSystemNamespace, ciliumPodK8s1, monitorLog)
+	result, targetIP := testPodConnectivityAcrossNodesAndReturnIP(kubectl, requireMultiNode, 2)
+	monitorStop()
+	ExpectWithOffset(1, result).Should(BeTrue(), "Connectivity test between nodes failed")
+
+	monitorPath := fmt.Sprintf("%s/%s", helpers.ReportDirectoryPath(), monitorLog)
+	By("Reading the monitor log at %s", monitorPath)
+	monitorOutput, err = ioutil.ReadFile(monitorPath)
+	ExpectWithOffset(1, err).To(BeNil(), "Could not read monitor log")
+	return monitorOutput, targetIP
+}
+
+func checkMonitorOutput(monitorOutput []byte, egressPktCount, ingressPktCount int) {
+	// Multiple connection attempts may be made, we need to
+	// narrow down to the last connection close, then match
+	// the ephemeral port + flags to ensure that the
+	// notifications match the table above.
+	egressTCPExpr := `TCP.*DstPort=80.*FIN=true`
+	egressTCPRegex := regexp.MustCompile(egressTCPExpr)
+	egressTCPMatches := egressTCPRegex.FindAll(monitorOutput, -1)
+	ExpectWithOffset(1, len(egressTCPMatches)).To(BeNumerically(">", 0), "Could not locate final FIN notification in monitor log")
+	finalMatch := egressTCPMatches[len(egressTCPMatches)-1]
+	portRegex := regexp.MustCompile(`SrcPort=([0-9]*)`)
+	// FindSubmatch should return ["SrcPort=12345" "12345"]
+	portBytes := portRegex.FindSubmatch(finalMatch)[1]
+
+	By("Looking for TCP notifications using the ephemeral port %q", portBytes)
+	port, err := strconv.Atoi(string(portBytes))
+	ExpectWithOffset(1, err).To(BeNil(), fmt.Sprintf("ephemeral port %q could not be converted to integer", string(portBytes)))
+	expEgress := fmt.Sprintf("SrcPort=%d", port)
+	expEgressRegex := regexp.MustCompile(expEgress)
+	egressMatches := expEgressRegex.FindAllIndex(monitorOutput, -1)
+	ExpectWithOffset(1, len(egressMatches)).To(Equal(egressPktCount), "Monitor log contained unexpected number of egress notifications matching %q", expEgress)
+	expIngress := fmt.Sprintf("DstPort=%d", port)
+	expIngressRegex := regexp.MustCompile(expIngress)
+	ingressMatches := expIngressRegex.FindAllIndex(monitorOutput, -1)
+	ExpectWithOffset(1, len(ingressMatches)).To(Equal(ingressPktCount), "Monitor log contained unexpected number of ingress notifications matching %q", expIngress)
 }

--- a/test/k8sT/DatapathConfiguration.go
+++ b/test/k8sT/DatapathConfiguration.go
@@ -97,6 +97,9 @@ var _ = Describe("K8sDatapathConfig", func() {
 				"--set global.bpf.monitorFlags=syn",
 				"--set global.debug.enabled=false",
 			})
+			// For local single-node testing, configure
+			// requireMultiNode to false and specify the node name.
+			requireMultiNode := true
 			nodeName := helpers.K8s1
 
 			var err error
@@ -105,7 +108,7 @@ var _ = Describe("K8sDatapathConfig", func() {
 
 			By(fmt.Sprintf("Launching cilium monitor on %q", ciliumPodK8s1))
 			monitorStop := kubectl.MonitorStart(helpers.KubeSystemNamespace, ciliumPodK8s1, monitorLog)
-			result, targetIP := testPodConnectivityAcrossNodesAndReturnIP(kubectl, 1)
+			result, targetIP := testPodConnectivityAcrossNodesAndReturnIP(kubectl, requireMultiNode, 1)
 			monitorStop()
 			cleanService()
 			Expect(result).Should(BeTrue(), "Connectivity test between nodes failed")
@@ -303,11 +306,11 @@ var _ = Describe("K8sDatapathConfig", func() {
 })
 
 func testPodConnectivityAcrossNodes(kubectl *helpers.Kubectl) bool {
-	result, _ := testPodConnectivityAcrossNodesAndReturnIP(kubectl, 1)
+	result, _ := testPodConnectivityAcrossNodesAndReturnIP(kubectl, true, 1)
 	return result
 }
 
-func fetchPodsWithOffset(kubectl *helpers.Kubectl, name, filter, hostIPAntiAffinity string, callOffset int) (targetPod string, targetPodJSON *helpers.CmdRes) {
+func fetchPodsWithOffset(kubectl *helpers.Kubectl, name, filter, hostIPAntiAffinity string, requireMultiNode bool, callOffset int) (targetPod string, targetPodJSON *helpers.CmdRes) {
 	callOffset++
 
 	// Fetch pod (names) with the specified filter
@@ -315,6 +318,10 @@ func fetchPodsWithOffset(kubectl *helpers.Kubectl, name, filter, hostIPAntiAffin
 	ExpectWithOffset(callOffset, err).Should(BeNil(), "Failure while waiting for connectivity test pods to start")
 	pods, err := kubectl.GetPodNames(helpers.DefaultNamespace, filter)
 	ExpectWithOffset(callOffset, err).Should(BeNil(), "Failure while retrieving pod name for %s", filter)
+	if requireMultiNode {
+		ExpectWithOffset(callOffset, len(pods)).Should(BeNumerically(">", 1),
+			fmt.Sprintf("This test requires at least two %s instances, but only one was found", name))
+	}
 
 	// Fetch the json description of one of the pods
 	targetPod = pods[0]
@@ -324,7 +331,7 @@ func fetchPodsWithOffset(kubectl *helpers.Kubectl, name, filter, hostIPAntiAffin
 
 	// If multinode / antiaffinity is required, ensure that the target is
 	// not on the same node as "hostIPAntiAffinity".
-	if hostIPAntiAffinity != "" {
+	if requireMultiNode && hostIPAntiAffinity != "" {
 		targetHost, err := targetPodJSON.Filter("{.status.hostIP}")
 		ExpectWithOffset(callOffset, err).Should(BeNil(), "Failure to retrieve host of pod %s", targetPod)
 
@@ -338,16 +345,16 @@ func fetchPodsWithOffset(kubectl *helpers.Kubectl, name, filter, hostIPAntiAffin
 	return targetPod, targetPodJSON
 }
 
-func testPodConnectivityAcrossNodesAndReturnIP(kubectl *helpers.Kubectl, callOffset int) (bool, string) {
+func testPodConnectivityAcrossNodesAndReturnIP(kubectl *helpers.Kubectl, requireMultiNode bool, callOffset int) (bool, string) {
 	callOffset++
 
 	By("Checking pod connectivity between nodes")
 
-	srcPod, srcPodJSON := fetchPodsWithOffset(kubectl, "client", "zgroup=testDSClient", "", callOffset)
+	srcPod, srcPodJSON := fetchPodsWithOffset(kubectl, "client", "zgroup=testDSClient", "", requireMultiNode, callOffset)
 	srcHost, err := srcPodJSON.Filter("{.status.hostIP}")
 	ExpectWithOffset(callOffset, err).Should(BeNil(), "Failure to retrieve host of pod %s", srcPod)
 
-	dstPod, dstPodJSON := fetchPodsWithOffset(kubectl, "server", "zgroup=testDS", srcHost.String(), callOffset)
+	dstPod, dstPodJSON := fetchPodsWithOffset(kubectl, "server", "zgroup=testDS", srcHost.String(), requireMultiNode, callOffset)
 	podIP, err := dstPodJSON.Filter("{.status.podIP}")
 	ExpectWithOffset(callOffset, err).Should(BeNil(), "Failure to retrieve IP of pod %s", dstPod)
 	targetIP := podIP.String()

--- a/test/k8sT/DatapathConfiguration.go
+++ b/test/k8sT/DatapathConfiguration.go
@@ -20,6 +20,7 @@ import (
 	"regexp"
 	"strconv"
 
+	"github.com/cilium/cilium/test/config"
 	. "github.com/cilium/cilium/test/ginkgo-ext"
 	"github.com/cilium/cilium/test/helpers"
 
@@ -66,14 +67,14 @@ var _ = Describe("K8sDatapathConfig", func() {
 	})
 
 	JustAfterEach(func() {
+		if !(config.CiliumTestConfig.HoldEnvironment && TestFailed()) {
+			// To avoid hitting GH-4384
+			kubectl.DeleteResource("service", "test-nodeport testds-service").ExpectSuccess(
+				"Service is deleted")
+		}
+
 		kubectl.ValidateNoErrorsInLogs(CurrentGinkgoTestDescription().Duration)
 	})
-
-	cleanService := func() {
-		// To avoid hit GH-4384
-		kubectl.DeleteResource("service", "test-nodeport testds-service").ExpectSuccess(
-			"Service is deleted")
-	}
 
 	deployCilium := func(options []string) {
 		DeployCiliumOptionsAndDNS(kubectl, options)
@@ -193,7 +194,6 @@ var _ = Describe("K8sDatapathConfig", func() {
 			})
 			validateBPFTunnelMap()
 			Expect(testPodConnectivityAcrossNodes(kubectl)).Should(BeTrue(), "Connectivity test with IPsec between nodes failed")
-			cleanService()
 		}, 600)
 
 		It("Check connectivity with sockops and VXLAN encapsulation", func() {
@@ -203,7 +203,6 @@ var _ = Describe("K8sDatapathConfig", func() {
 			})
 			validateBPFTunnelMap()
 			Expect(testPodConnectivityAcrossNodes(kubectl)).Should(BeTrue(), "Connectivity test between nodes failed")
-			cleanService()
 		}, 600)
 
 		It("Check connectivity with VXLAN encapsulation", func() {
@@ -212,7 +211,6 @@ var _ = Describe("K8sDatapathConfig", func() {
 			})
 			validateBPFTunnelMap()
 			Expect(testPodConnectivityAcrossNodes(kubectl)).Should(BeTrue(), "Connectivity test between nodes failed")
-			cleanService()
 		}, 600)
 
 		It("Check connectivity with Geneve encapsulation", func() {
@@ -221,7 +219,6 @@ var _ = Describe("K8sDatapathConfig", func() {
 			})
 			validateBPFTunnelMap()
 			Expect(testPodConnectivityAcrossNodes(kubectl)).Should(BeTrue(), "Connectivity test between nodes failed")
-			cleanService()
 		})
 
 		It("Check vxlan connectivity with per endpoint routes", func() {
@@ -231,7 +228,6 @@ var _ = Describe("K8sDatapathConfig", func() {
 				"--set global.autoDirectNodeRoutes=true",
 			})
 			Expect(testPodConnectivityAcrossNodes(kubectl)).Should(BeTrue(), "Connectivity test between nodes failed")
-			cleanService()
 		})
 
 	})
@@ -247,7 +243,6 @@ var _ = Describe("K8sDatapathConfig", func() {
 
 			deployCilium(directRoutingOptions)
 			Expect(testPodConnectivityAcrossNodes(kubectl)).Should(BeTrue(), "Connectivity test between nodes failed")
-			cleanService()
 		})
 
 		It("Check direct connectivity with per endpoint routes", func() {
@@ -257,7 +252,6 @@ var _ = Describe("K8sDatapathConfig", func() {
 				"--set global.endpointRoutes.enabled=true",
 			))
 			Expect(testPodConnectivityAcrossNodes(kubectl)).Should(BeTrue(), "Connectivity test between nodes failed")
-			cleanService()
 		})
 	})
 
@@ -272,7 +266,6 @@ var _ = Describe("K8sDatapathConfig", func() {
 				"--set global.encryption.interface=enp0s8",
 			})
 			Expect(testPodConnectivityAcrossNodes(kubectl)).Should(BeTrue(), "Connectivity test between nodes failed")
-			cleanService()
 		})
 	})
 
@@ -286,7 +279,6 @@ var _ = Describe("K8sDatapathConfig", func() {
 				"--set global.ipv6.enabled=false",
 			})
 			Expect(testPodConnectivityAcrossNodes(kubectl)).Should(BeTrue(), "Connectivity test between nodes failed")
-			cleanService()
 		})
 	})
 
@@ -300,7 +292,6 @@ var _ = Describe("K8sDatapathConfig", func() {
 				"--set global.etcd.managed=true",
 			})
 			Expect(testPodConnectivityAcrossNodes(kubectl)).Should(BeTrue(), "Connectivity test between nodes failed")
-			cleanService()
 		})
 	})
 })

